cpg_pipes/targets.py

```diff
@@ -0,0 +1,640 @@
+    MALE = 1
+    FEMALE = 2
+
+    @staticmethod
+    def parse(sex: str | None) -> 'Sex':
+        """
+        Parse a string into a Sex object.
+        """
+        if sex:
+            if sex.lower() in ('m', 'male', '1'):
+                return Sex.MALE
+            if sex.lower() in ('f', 'female', '2'):
+                return Sex.FEMALE
+        return Sex.UNKNOWN
+
+
+class Sample(Target):
+    """
+    Represents a Sample.
+    """
+
+    def __init__(
+        self,
+        id: str,  # pylint: disable=redefined-builtin
+        dataset: 'Dataset',  # type: ignore  # noqa: F821
+        external_id: str | None = None,
+        sequencing_type: SequencingType | None = None,
+        participant_id: str | None = None,
+        meta: dict | None = None,
+        sex: Sex | None = None,
+        pedigree: Optional['PedigreeInfo'] = None,
+        alignment_input: AlignmentInput | None = None,
+    ):
+        super().__init__()
+        self.id = id
+        self._external_id = external_id
+        self.dataset = dataset
+        self.sequencing_type = sequencing_type or SequencingType.GENOME
+        self._participant_id = participant_id
+        self.meta: dict = meta or dict()
+        self.pedigree: PedigreeInfo | None = pedigree
+        if sex:
+            self.pedigree = PedigreeInfo(
+                sample=self,
+                fam_id=self.participant_id,
+                sex=sex,
+            )
+        self.alignment_input: AlignmentInput | None = alignment_input
+
+    def __repr__(self):
+        return (
+            f'Sample({self.dataset.name}/{self.id}'
+            + (f'|{self._external_id}' if self._external_id else '')
+            + (f', participant={self._participant_id}' if self._participant_id else '')
+            + f', forced={self.forced}'
+            + f', active={self.active}'
+            + f', meta={self.meta}'
+            + f', sequencing_type={self.sequencing_type.value}'
+            + (
+                f', alignment_input={self.alignment_input}'
+                if self.alignment_input
+                else ''
+            )
+            + (f', pedigree={self.pedigree}' if self.pedigree else '')
+            + f')'
+        )
+
+    def __str__(self):
+        ai_tag = ''
+        if self.alignment_input:
+            if isinstance(self.alignment_input, CramPath):
+                if self.alignment_input.is_bam:
+                    ai_tag = f'|SEQ=CRAM'
+                else:
+                    ai_tag = f'|SEQ=BAM'
+            else:
+                ai_tag = f'|SEQ={len(self.alignment_input)}FQS'
+
+        ext_id = f'|{self._external_id}' if self._external_id else ''
+        return f'Sample({self.dataset.name}/{self.id}{ext_id}{ai_tag})'
+
+    @property
+    def participant_id(self) -> str:
+        """
+        Get ID of participant corresponding to this sample,
+        or substitute it with external ID.
+        """
+        return self._participant_id or self.external_id
```

Comment:
Should all samples that go through this process have a participant_id, should this be an assert?

Reply:
I don't want to force non-CPG users to use participant IDs. I think they should be able to use only one type of IDs if they want, and the code that expects participant IDs still should work